### ISSUE TITLE
Test failures

### ISSUE THREAD

**Issue by jaraco** [pypa/setuptools-scm#9]
==================================== ERRORS ====================================
__________________________ ERROR at setup of test_do ___________________________

```
@pytest.mark.parametrize('cmd', ['ls', 'dir'])
def test_do(cmd, tmpdir):
    LookupError: no factory found for function argument 'cmd'
    available funcargs: pytestconfig, capsys, capfd, tmpdir, monkeypatch, recwarn, case, data, expected
    use 'py.test --funcargs [testpath]' for help on them.
```

/var/tmp/portage/dev-python/hgdistver-0.16/work/hgdistver-0.16/test_hgdistver.py:22
__________________ ERROR at setup of test_archival_to_version __________________

request = <FuncargRequest for <Function 'test_archival_to_version'>>

>   def pytest_funcarg__expected(request): return request.param
> E   AttributeError: FuncargRequest instance has no attribute 'param'

test_hgdistver.py:59: AttributeError
__________________ ERROR at setup of test_version_from_hg_id ___________________

```
@py.test.mark.parametrize('method', ['version_from_hg15_parents', 'version_from_hg_log_with_tags'])
def test_version_from_hg_id(tmpdir, method):
    LookupError: no factory found for function argument 'method'
    available funcargs: pytestconfig, capsys, capfd, tmpdir, monkeypatch, recwarn, case, data, expected
    use 'py.test --funcargs [testpath]' for help on them.
```

/var/tmp/portage/dev-python/hgdistver-0.16/work/hgdistver-0.16/test_hgdistver.py:95
=================================== FAILURES ===================================
____________________________ test_version_from_git _____________________________

tmpdir = local('/var/tmp/portage/dev-python/hgdistver-0.16/temp/pytest-0/test_version_from_git0')

```
def test_version_from_git(tmpdir):
    cwd = str(tmpdir)
    do('git init', cwd)
    initial = get_version(cwd)
    assert initial == '0.0'
    tmpdir.join('test.txt').write('test')
    do('git add test.txt', cwd)
    do('git commit -m commit', cwd)
    after_first_commit = get_version(cwd)
```

> ```
>   assert after_first_commit.startswith('0.0.post1-')
> ```
> 
> E       assert False
> E        +  where False = <built-in method startswith of str object at 0x17ff2d8>('0.0.post1-')
> E        +    where <built-in method startswith of str object at 0x17ff2d8> = '0.0'.startswith

test_hgdistver.py:77: AssertionError
------------------------------- Captured stdout --------------------------------
cmd 'git init'
out 'Initialized empty Git repository in /var/tmp/portage/dev-python/hgdistver-0.16/temp/pytest-0/test_version_from_git0/.git/\n'
cmd 'git rev-parse --verify --quiet HEAD'
ret 1
cmd 'git add test.txt'
cmd 'git commit -m commit'
err '\n**\* Please tell me who you are.\n\nRun\n\n  git config --global user.email "you@example.com"\n  git config --global user.name "Your Name"\n\nto set your account\'s default identity.\nOmit --global to set the identity only in this repository.\n\nfatal: empty ident  <djc@miles.(none)> not allowed\n'
ret 128
**\* Please tell me who you are.

Run

  git config --global user.email "you@example.com"
  git config --global user.name "Your Name"

to set your account's default identity.
Omit --global to set the identity only in this repository.

fatal: empty ident  <djc@miles.(none)> not allowed
cmd 'git rev-parse --verify --quiet HEAD'
ret 1
================= 1 failed, 8 passed, 3 error in 0.20 seconds ==================

---
- Bitbucket: https://bitbucket.org/pypa/setuptools_scm/issue/9
- Originally reported by: Dirkjan Ochtman
- Originally created at: 2012-10-10T10:04:31.469

**Comment by jaraco** [comment]
Also, you should declare a test_requires pytest and compatibly Python versions in the setup.py.

---

Original comment by: Dirkjan Ochtman

**Comment by jaraco** [comment]
(Reply via Ronn...@gmx.de):

What version of pytest?

also maybe notify holger in the pytest issue tracker

---

Original comment by: Ronny Pfannschmidt

**Comment by jaraco** [comment]
eeeeer,  progress?
with py.test

```
 * python3_2: running distutils-r1_run_phase python_test
========================================================= test session starts =========================================================
platform linux2 -- Python 3.2.4 -- pytest-2.3.4
INTERNALERROR> Traceback (most recent call last):
INTERNALERROR>   File "/usr/lib64/python3.2/site-packages/_pytest/main.py", line 81, in wrap_session
INTERNALERROR>     doit(config, session)
INTERNALERROR>   File "/usr/lib64/python3.2/site-packages/_pytest/main.py", line 112, in _main
INTERNALERROR>     config.hook.pytest_collection(session=session)
INTERNALERROR>   File "/usr/lib64/python3.2/site-packages/_pytest/core.py", line 422, in __call__
INTERNALERROR>     return self._docall(methods, kwargs)
INTERNALERROR>   File "/usr/lib64/python3.2/site-packages/_pytest/core.py", line 433, in _docall
INTERNALERROR>     res = mc.execute()
INTERNALERROR>   File "/usr/lib64/python3.2/site-packages/_pytest/core.py", line 351, in execute
INTERNALERROR>     res = method(**kwargs)
INTERNALERROR>   File "/usr/lib64/python3.2/site-packages/_pytest/main.py", line 116, in pytest_collection
INTERNALERROR>     return session.perform_collect()
INTERNALERROR>   File "/usr/lib64/python3.2/site-packages/_pytest/main.py", line 467, in perform_collect
INTERNALERROR>     config=self.config, items=items)
INTERNALERROR>   File "/usr/lib64/python3.2/site-packages/_pytest/core.py", line 422, in __call__
INTERNALERROR>     return self._docall(methods, kwargs)
INTERNALERROR>   File "/usr/lib64/python3.2/site-packages/_pytest/core.py", line 433, in _docall
INTERNALERROR>     res = mc.execute()
INTERNALERROR>   File "/usr/lib64/python3.2/site-packages/_pytest/core.py", line 351, in execute
INTERNALERROR>     res = method(**kwargs)
INTERNALERROR>   File "/usr/lib64/python3.2/site-packages/_pytest/python.py", line 1511, in pytest_collection_modifyitems
INTERNALERROR>     items[:] = parametrize_sorted(items, set(), {}, 0)
INTERNALERROR>   File "/usr/lib64/python3.2/site-packages/_pytest/python.py", line 1698, in parametrize_sorted
INTERNALERROR>     argparamlist = getfuncargparams(item, ignore, scopenum, cache)
INTERNALERROR>   File "/usr/lib64/python3.2/site-packages/_pytest/python.py", line 1726, in getfuncargparams
INTERNALERROR>     argparams = [x for x in cs.params.items() if x not in ignore
INTERNALERROR>   File "/usr/lib64/python3.2/site-packages/_pytest/python.py", line 1726, in <listcomp>
INTERNALERROR>     argparams = [x for x in cs.params.items() if x not in ignore
INTERNALERROR> TypeError: unhashable type: 'dict'

==========================================================  in 0.06 seconds ===========================================================
 * ERROR: dev-python/hgdistver-0.16-r1 failed (test phase):
 *   Tests failed under python3.2
```

nosetests;

```

 * python3_2: running distutils-r1_run_phase python_test
EEEEEEEEEEEE
======================================================================
ERROR: test_hgdistver.test_do
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib64/python3.2/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
TypeError: test_do() takes exactly 2 arguments (0 given)

======================================================================
ERROR: test_hgdistver.test_data_from_archival
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib64/python3.2/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
TypeError: test_data_from_archival() takes exactly 1 argument (0 given)

======================================================================
ERROR: test_hgdistver.test_archival_to_version
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib64/python3.2/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
TypeError: test_archival_to_version() takes exactly 2 arguments (0 given)

======================================================================
ERROR: test_hgdistver.test_version_from_git
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib64/python3.2/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
TypeError: test_version_from_git() takes exactly 1 argument (0 given)

======================================================================
ERROR: test_hgdistver.test_version_from_hg_id
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib64/python3.2/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
TypeError: test_version_from_hg_id() takes exactly 2 arguments (0 given)

======================================================================
ERROR: test_hgdistver.test_version_from_archival
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib64/python3.2/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
TypeError: test_version_from_archival() takes exactly 1 argument (0 given)

======================================================================
ERROR: test_hgdistver.test_version_from_cachefile
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib64/python3.2/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
TypeError: test_version_from_cachefile() takes exactly 1 argument (0 given)

======================================================================
ERROR: test_hgdistver.test_version_from_pkginfo
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib64/python3.2/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
TypeError: test_version_from_pkginfo() takes exactly 1 argument (0 given)

======================================================================
ERROR: test_hgdistver.test_root_parameter_creation
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib64/python3.2/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
TypeError: test_root_parameter_creation() takes exactly 1 argument (0 given)

======================================================================
ERROR: test_hgdistver.test_root_parameter_pass_by
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib64/python3.2/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
TypeError: test_root_parameter_pass_by() takes exactly 1 argument (0 given)

======================================================================
ERROR: test_hgdistver.test_cachefile_join
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib64/python3.2/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
TypeError: test_cachefile_join() takes exactly 1 argument (0 given)

======================================================================
ERROR: test_hgdistver.test_recreate_cachefile_from_pkginfo
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib64/python3.2/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
TypeError: test_recreate_cachefile_from_pkginfo() takes exactly 1 argument (0 given)

----------------------------------------------------------------------
Ran 12 tests in 0.059s

FAILED (errors=12)
 * ERROR: dev-python/hgdistver-0.16-r1 failed (test phase):
```

12 out of 12. perfect, eer, score
train wreck

---

Original comment by: [idella5](http://bitbucket.org/idella5)

**Comment by jaraco** [comment]
grabing it now

---

Original comment by: Ronny Pfannschmidt

**Comment by jaraco** [comment]
uploaded a fixed release, please test

---

Original comment by: Ronny Pfannschmidt

**Comment by jaraco** [comment]
With py.test, I still see these failures with hgdistver-0.17 (both 2.7 and 3.2):

```
#!python

=================================== FAILURES ===================================
____________________________ test_version_from_git _____________________________

tmpdir = local('/var/tmp/portage/dev-python/hgdistver-0.17/temp/python2.7/pytest-0/test_version_from_git0')

    def test_version_from_git(tmpdir):
        cwd = str(tmpdir)
        do('git init', cwd)
        initial = get_version(cwd)
        assert initial == '0.0'
        tmpdir.join('test.txt').write('test')
        do('git add test.txt', cwd)
        do('git commit -m commit', cwd)
        after_first_commit = get_version(cwd)
    .
>       assert after_first_commit.startswith('0.0.post1-')
E       assert <built-in method startswith of str object at 0x2d38d00>('0.0.post1-')
E        +  where <built-in method startswith of str object at 0x2d38d00> = '0.0'.startswith

=================================== FAILURES ===================================
____________________________ test_version_from_git _____________________________

tmpdir = local('/var/tmp/portage/dev-python/hgdistver-0.17/temp/python3.2/pytest-0/test_version_from_git0')

    def test_version_from_git(tmpdir):
        cwd = str(tmpdir)
        do('git init', cwd)
        initial = get_version(cwd)
        assert initial == '0.0'
        tmpdir.join('test.txt').write('test')
        do('git add test.txt', cwd)
        do('git commit -m commit', cwd)
        after_first_commit = get_version(cwd)

>       assert after_first_commit.startswith('0.0.post1-')
E       assert <built-in method startswith of str object at 0x109cfb8>('0.0.post1-')
E        +  where <built-in method startswith of str object at 0x109cfb8> = '0.0'.startswith

test_hgdistver.py:84: AssertionError
------------------------------- Captured stdout --------------------------------
cmd 'git init'
out b'Initialized empty Git repository in /var/tmp/portage/dev-python/hgdistver-0.17/temp/python3.2/pytest-0/test_version_from_git0/.git/\n'
cmd 'git rev-parse --verify --quiet HEAD'
ret 1
cmd 'git add test.txt'
cmd 'git commit -m commit'
err b'\n*** Please tell me who you are.\n\nRun\n\n  git config --global user.email "you@example.com"\n  git config --global user.name "Your Name"\n\nto set your account\'s default identity.\nOmit --global to set the identity only in this repository.\n\nfatal: empty ident name (for <djc@miles.(none)>) not allowed\n'
ret 128
*** Please tell me who you are.

Run

  git config --global user.email "you@example.com"
  git config --global user.name "Your Name"

to set your account's default identity.
Omit --global to set the identity only in this repository.

fatal: empty ident name (for <djc@miles.(none)>) not allowed
cmd 'git rev-parse --verify --quiet HEAD'
ret 1
____________ test_version_from_hg_id[version_from_hg_log_with_tags] ____________

tmpdir = local('/var/tmp/portage/dev-python/hgdistver-0.17/temp/python3.2/pytest-0/test_version_from_hg_id_version_from_hg_log_with_tags_0')
method = 'version_from_hg_log_with_tags'

    @py.test.mark.parametrize('method', [
        'version_from_hg15_parents',
        'version_from_hg_log_with_tags'])
    def test_version_from_hg_id(tmpdir, method):
        hgv = _hg_version()
        print(hgv)
        if hgv < '1.5' and 'parents' in method:
            py.test.skip('hg too old, this test needs >=1.5')
        cwd = str(tmpdir)
        do('hg init', cwd)
        initial = get_version(cwd, method=method)
>       assert initial == '0.0'
E       assert '0.0.post1-000000000000' == '0.0'
E         - 0.0.post1-000000000000
E         + 0.0

test_hgdistver.py:112: AssertionError
------------------------------- Captured stdout --------------------------------
cmd 'hg --version'
out b'Mercurial Distributed SCM (version 2.6.2)\n(see http://mercurial.selenic.com for more information)\n\nCopyright (C) 2005-2012 Matt Mackall and others\nThis is free software; see the source for copying conditions. There is NO\nwarranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.\n'
2.6.2
cmd 'hg init'
cmd 'hg id -i'
out b'000000000000\n'
===================== 2 failed, 16 passed in 4.30 seconds ======================


test_hgdistver.py:84: AssertionError
------------------------------- Captured stdout --------------------------------
cmd 'git init'
out 'Initialized empty Git repository in /var/tmp/portage/dev-python/hgdistver-0.17/temp/python2.7/pytest-0/test_version_from_git0/.git/\n'
cmd 'git rev-parse --verify --quiet HEAD'
ret 1
cmd 'git add test.txt'
cmd 'git commit -m commit'
err '\n*** Please tell me who you are.\n\nRun\n\n  git config --global user.email "you@example.com"\n  git config --global user.name "Your Name"\n\nto set your account\'s default identity.\nOmit --global to set the identity only in this repository.\n\nfatal: empty ident name (for <djc@miles.(none)>) not allowed\n'
ret 128
*** Please tell me who you are.

Run

  git config --global user.email "you@example.com"
  git config --global user.name "Your Name"

to set your account's default identity.
Omit --global to set the identity only in this repository.

fatal: empty ident name (for <djc@miles.(none)>) not allowed
cmd 'git rev-parse --verify --quiet HEAD'
ret 1
____________ test_version_from_hg_id[version_from_hg_log_with_tags] ____________

tmpdir = local('/var/tmp/portage/dev-python/hgdistver-0.17/temp/python2.7/pytest-0/test_version_from_hg_id_version_from_hg_log_with_tags_0')
method = 'version_from_hg_log_with_tags'

    @py.test.mark.parametrize('method', [
        'version_from_hg15_parents',
        'version_from_hg_log_with_tags'])
    def test_version_from_hg_id(tmpdir, method):
        hgv = _hg_version()
        print(hgv)
        if hgv < '1.5' and 'parents' in method:
            py.test.skip('hg too old, this test needs >=1.5')
        cwd = str(tmpdir)
        do('hg init', cwd)
        initial = get_version(cwd, method=method)
>       assert initial == '0.0'
E       assert u'0.0.post1-000000000000' == '0.0'
E         - 0.0.post1-000000000000
E         + 0.0

test_hgdistver.py:112: AssertionError
------------------------------- Captured stdout --------------------------------
cmd 'hg --version'
out 'Mercurial Distributed SCM (version 2.6.2)\n(see http://mercurial.selenic.com for more information)\n\nCopyright (C) 2005-2012 Matt Mackall and others\nThis is free software; see the source for copying conditions. There is NO\nwarranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.\n'
2.6.2
cmd 'hg init'
cmd 'hg id -i'
out '000000000000\n'
===================== 2 failed, 16 passed in 4.35 seconds ======================
```

---

Original comment by: Dirkjan Ochtman

**Comment by jaraco** [comment]
seems there are some version differences with never versions that i need to take into account

---

Original comment by: Ronny Pfannschmidt

**Comment by jaraco** [comment]
Ping, please take a look at this.

---

Original comment by: Dirkjan Ochtman

**Comment by jaraco** [comment]
i think it should work now, cant be sure tho

---

Original comment by: Ronny Pfannschmidt

**Comment by jaraco** [comment]
Hmm, not quite. With 0.17:

```
#!python

============================================= FAILURES =============================================
______________________________________ test_version_from_git _______________________________________

tmpdir = local('/var/tmp/portage/dev-python/hgdistver-0.17/temp/python2.7/pytest-0/test_version_from_git0')

    def test_version_from_git(tmpdir):
        cwd = str(tmpdir)
        do('git init', cwd)
        initial = get_version(cwd)
        assert initial == '0.0'
        tmpdir.join('test.txt').write('test')
        do('git add test.txt', cwd)
        do('git commit -m commit', cwd)
        after_first_commit = get_version(cwd)

>       assert after_first_commit.startswith('0.0.post1-')
E       assert <built-in method startswith of str object at 0x1ca5af8>('0.0.post1-')
E        +  where <built-in method startswith of str object at 0x1ca5af8> = '0.0'.startswith

test_hgdistver.py:84: AssertionError
----------------------------------------- Captured stdout ------------------------------------------
cmd 'git init'
out 'Initialized empty Git repository in /var/tmp/portage/dev-python/hgdistver-0.17/temp/python2.7/pytest-0/test_version_from_git0/.git/\n'
cmd 'git rev-parse --verify --quiet HEAD'
ret 1
cmd 'git add test.txt'
cmd 'git commit -m commit'
err '\n*** Please tell me who you are.\n\nRun\n\n  git config --global user.email "you@example.com"\n  git config --global user.name "Your Name"\n\nto set your account\'s default identity.\nOmit --global to set the identity only in this repository.\n\nfatal: empty ident name (for <djc@miles.(none)>) not allowed\n'
ret 128
*** Please tell me who you are.

Run

  git config --global user.email "you@example.com"
  git config --global user.name "Your Name"

to set your account's default identity.
Omit --global to set the identity only in this repository.

fatal: empty ident name (for <djc@miles.(none)>) not allowed
cmd 'git rev-parse --verify --quiet HEAD'
ret 1
______________________ test_version_from_hg_id[version_from_hg_log_with_tags] ______________________

tmpdir = local('/var/tmp/portage/dev-python/hgdistver-0.17/temp/python2.7/pytest-0/test_version_from_hg_id_version_from_hg_log_with_tags_0')
method = 'version_from_hg_log_with_tags'

    @py.test.mark.parametrize('method', [
        'version_from_hg15_parents',
        'version_from_hg_log_with_tags'])
    def test_version_from_hg_id(tmpdir, method):
        hgv = _hg_version()
        print(hgv)
        if hgv < '1.5' and 'parents' in method:
            py.test.skip('hg too old, this test needs >=1.5')
        cwd = str(tmpdir)
        do('hg init', cwd)
        initial = get_version(cwd, method=method)
>       assert initial == '0.0'
E       assert u'0.0.post1-000000000000' == '0.0'
E         - 0.0.post1-000000000000
E         + 0.0

test_hgdistver.py:112: AssertionError
----------------------------------------- Captured stdout ------------------------------------------
cmd 'hg --version'
out 'Mercurial Distributed SCM (version 2.9.1)\n(see http://mercurial.selenic.com for more information)\n\nCopyright (C) 2005-2014 Matt Mackall and others\nThis is free software; see the source for copying conditions. There is NO\nwarranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.\n'
2.9.1
cmd 'hg init'
cmd 'hg id -i'
out '000000000000\n'
=============================== 2 failed, 16 passed in 4.38 seconds ================================
```

With 0.18:

```
#!python

======================================= test session starts ========================================
platform linux -- Python 3.3.3 -- pytest-2.3.5
collected 0 items

=========================================  in 0.02 seconds =========================================
======================================= test session starts ========================================
platform linux2 -- Python 2.7.5 -- pytest-2.3.5
collected 0 items

=========================================  in 0.01 seconds =========================================
```

---

Original comment by: Dirkjan Ochtman

**Comment by jaraco** [comment]
its fixed in 0.19 

---

Original comment by: Ronny Pfannschmidt

**Comment by jaraco** [comment]
Okay, I'll take a look when that's released.

---

Original comment by: Dirkjan Ochtman

**Comment by jaraco** [comment]
It's also fixed in  0.18

---

Original comment by: [idella5](http://bitbucket.org/idella5)

**Comment by jaraco** [comment]
should be fixed, please reopen if it happens a gain (git is a flaky mess)

---

Original comment by: Ronny Pfannschmidt
